runtime/src/tasks/fifo/tasks-fifo.c

```diff
@@ -389,8 +394,81 @@ void chpl_task_exit(void) {
 }
 
 
-void chpl_task_callMain(void (*chpl_main)(void)) {
+typedef void (*main_ptr_t)(void);
+static void* do_callMain(void* arg) {
+  main_ptr_t chpl_main = (main_ptr_t) arg;
+
+  // make sure this thread has thread-private data.
+  setup_main_thread_private_data();
+
   chpl_main();
+  return NULL;
+}
+
+void chpl_task_callMain(void (*chpl_main)(void)) {
+  // since we want to run all work in a task with a comm-friendly stack,
+  // run main in a pthread that we will wait for.
+  size_t stack_size;
+  pthread_attr_t attr;
+  pthread_t thread;
+  int rc;
+#ifndef CHPL_USING_CSTDLIB_MALLOC
+  size_t align;
+  int free_flag;
+  void* stack;
+#endif
+
+  rc = pthread_attr_init(&attr);
+  if( rc != 0 ) {
+    chpl_internal_error("pthread_attr_init main failed");
+  }
+
+  stack_size  = chpl_thread_getCallStackSize();
+
+/* Uses a registered stack, if possible. For the reason about
+ * this define, see threads-pthreads.c
+ */
+#ifndef CHPL_USING_CSTDLIB_MALLOC
+  align = chpl_getSysPageSize();
+  stack = chpl_memalign(align, stack_size + align);
+  if( stack == NULL ){
+    chpl_internal_error("chpl_memalign main failed");
+  }
+
+  rc = mprotect((unsigned char *)stack + stack_size, align, PROT_NONE);
```

Comment:
I've noticed the second problem while I was answering the first question before and I've already correct it (I'm testing it at the moment). 

For the first one: ok, I' ll follow your suggestion.